File: clooey/clooey.py
import ast
from dataclasses import dataclass
from jinja2 import Template
from typing import List
import re


@dataclass
class Input:
    label: str
    placeholder: str = ''


@dataclass
class CLI:
    inputs: List[Input]
    title: str = ''
    description: str = ''


class InputCallVisitor(ast.NodeVisitor):
    def __init__(self):
        self.input_calls = []

    def visit_Call(self, node):
        # Check if the function called is 'input'
        if isinstance(node.func, ast.Name) and node.func.id == 'input':
            self.input_calls.append(parse_prompt(node.args[0].value)) # input only takes one arg
        self.generic_visit(node)


def parse_prompt(prompt):
    """Parses a prompt string into a dictionary with 'label' and 'placeholder' keys.
    
    Args:
        prompt (str): The input string to parse.
    
    >>> parse_prompt("Your city [Seattle]:")
    Input(label='Your city:', placeholder='Seattle')
    
    >>> parse_prompt("Your city: [Seattle]")
    Input(label='Your city:', placeholder='Seattle')
    
    >>> parse_prompt("Your city \\[Seattle\\]:")
    Input(label='Your city [Seattle]:', placeholder='')
    
    >>> parse_prompt("Your city \\[Seattle\\]: [Placeholder]")
    Input(label='Your city [Seattle]:', placeholder='Placeholder')
    """
    pattern = r'^(.*)?(?<!\\)\[(.*)(?<!\\)\](.*)?$'
    match = re.match(pattern, prompt)
    
    if match:
        label = match.group(1).strip() + match.group(3).strip()
        placeholder = match.group(2).strip() if match.group(2) else ''
    else:
        label, placeholder = prompt, ''
    
    label = label.replace('\\[', '[').replace('\\]', ']')
    return Input(label, placeholder)


def parse(program):
    """Parse a program, and return a representation of its CLI
    
    Args:
        program (str): Stringified code to parse

    Returns:
        List[Input]: Specifying CLI using a list of Input objects

    >>> cli = parse(\"'''\\nBlurple\\nBlack and purple color mix\\n'''\")
    >>> cli.title
    'Blurple'
    >>> cli.description
    'Black and purple color mix'
    >>> cli = parse(\"'''Blurple\\nBlack and purple color mix\\n'''\")
    >>> cli.title
    'Blurple'
    >>> cli = parse(\"'''Blurple\\n\\nBlack and purple color mix\\n'''\")
    >>> cli.title
    'Blurple'
    """
    tree = ast.parse(program)
    visitor = InputCallVisitor()
    visitor.visit(tree)
    
    if docs := ast.get_docstring(tree):
        title, description = docs.strip().split('\n', 1)
    return CLI(visitor.input_calls, title.strip(), description.strip())


TEMPLATE_FORM = """
{% if cli.title %}<h1>{{ cli.title }}</h1>{% endif %}
{% if cli.description %}<p>{{ cli.description }}</p>{% endif %}
<form method="post">
    {% for input in cli.inputs %}
    <label>{{ input.label }}</label>
    <input type="text" name="{{ loop.index }}" placeholder="{{ input.placeholder }}">
    {% endfor %}
    <input type="submit" value="submit">
</form>
""".strip()


def generate(cli, template):
    """Take a CLI, and generate a basic form.
    
    Args:
        cli List[Input]: Returned representation of CLI from parse
        template str: Jinja2 template for form HTML
    """
    return Template(template).render(cli=cli)


def execute(responses, program):
    """
    Executes the program using a list of provided form submission.
    
    Args:
        responses List[str]: Responses for each input call prompt
        program (str): Stringified code to run
    """
    inputs_iterator = iter(responses)

    def custom_input(prompt=None):
        """
        Custom implementation of input that simply returns from the iterable of
        responses. In this way, we can mimic user interaction with the script.
        """
        try:
            return next(inputs_iterator)
        except StopIteration:
            raise RuntimeError("Not enough inputs provided")

    exec(program, {'input': custom_input})

File: js/clooey.js
class CLI {
    constructor(inputs, title = '', description = '') {
        this.inputs = inputs;
        this.title = title;
        this.description = description;
    }
}

class Input {
    constructor(label, placeholder = '') {
        this.label = label;
        this.placeholder = placeholder;
    }
}

/**
 * Converts a piece of Python source code into a list of prompts. Assumes that
 * all calls to the input function contain only a string as an argument. No
 * variables.
 * 
 * @param {*} pythonCode - stringified Python source code
 * @returns - list of strings, representing input prompts
 */
function parse(pythonCode) {
    const inputPromptRegex = /input\(["']([^"']+)["']\)/g;
    let matches;
    const prompts = [];

    while ((matches = inputPromptRegex.exec(pythonCode)) !== null) {
        prompts.push(parsePrompt(matches[1]));
    }

    const docComment = pythonCode.match(/(?:\"\"\"|\'\'\')([\s\S]*?)(?:\"\"\"|\'\'\')/);
    let title = '', description = '';
    
    if (docComment) {
        const docs = docComment[1].trim().split('\n');
        title = docs[0].trim();
        description = docs.slice(1).join('\n').trim();
    }

    return new CLI(prompts, title, description);
}

/**
 * Parse the prompt into a label and placeholder.
 * 
 * @param {*} prompt - raw string prompt
 * @returns - Input object
 * 
 * Examples:
 * 
 * console.log(parsePrompt("Your city [Seattle]:"));
 * console.log(parsePrompt("Your city: [Seattle]"));
 * console.log(parsePrompt("Your city \\[Seattle\\]:"));
 * console.log(parsePrompt("Your city \\[Seattle\\]: [Placeholder]"));
 * 
 * should give the following outputs
 * 
 * Input {label: 'Your city :', placeholder: 'Seattle'}
 * Input {label: 'Your city:', placeholder: 'Seattle'}
 * Input {label: 'Your city [Seattle]:', placeholder: ''}
 * Input {label: 'Your city [Seattle]:', placeholder: 'Placeholder'}
 */
function parsePrompt(prompt) {
    // Define the pattern to match the input string
    const pattern = /^(.*?)(?<!\\)\[(.*?)(?<!\\)\](.*)?$/;
    const match = prompt.match(pattern);
    
    let label, placeholder;
    
    if (match) {
        var suffix = match[3] ? match[3] : '';
        label = (match[1] + suffix).trim();
        placeholder = match[2].trim();
    } else {
        label = prompt;
        placeholder = '';
    }

    // Replace escaped square brackets with actual square brackets
    label = label.replace('\\[', '[').replace('\\]', ']');
    
    return new Input(label, placeholder);
}

TEMPLATE_FORM = `
{% if cli.title %}<h1>{{ cli.title }}</h1>{% endif %}
{% if cli.description %}<p>{{ cli.description }}</p>{% endif %}
<form method="post">
    {% for input in cli.inputs %}
    <label>{{ input.label }}</label>
    <input type="text" name="{{ loop.index }}" placeholder="{{ input.placeholder }}">
    {% endfor %}
    <input type="submit" value="submit">
</form>
`

/**
 * Generates form from CLI.
 * 
 * Note that `nunjucks does not sandbox execution so it is not safe to run
 * user-defined templates or inject user-defined content into template
 * definitions.` As a result, the template passed to this function should be
 * developer-provided, not user-provided.
 * 
 * @param {*} cli - list of strings, representings input prompts
 * @param {*} template - nunjucks template
 * @returns - HTML form
 */
function generate(cli, template) {
    template = template || TEMPLATE_FORM;
    return nunjucks.renderString(template, { cli });
}

/**
 * Execute the provided Python script with the given predetermined input
 * responses.
 * 
 * TODO: May need to save and restore the original stdin handler.
 * 
 * @param {*} responses - list of strings, representing responses
 * @param {*} pythonCode - stringified Python source code
 * @param {*} pyodide - optional argument, for the active pyodide instance
 */
async function execute(responses, pythonCode, pyodide) {
    if (!pyodide) {
        const { loadPyodide } = await import("https://cdn.jsdelivr.net/pyodide/v0.26.1/full/pyodide.mjs");
        pyodide = await loadPyodide();
    }

    // Calls to stdin simply returns from the list of responses
    var responseIndex = 0;
    pyodide.setStdin({ stdin: () => { return responses[responseIndex++ ]; }});

    // Run Python
    var out = pyodide.runPython(pythonCode);
    return out;
}

/**
 * Convenience utility to get submission data from a form.
 * 
 * @param {*} el - DOM element
 * @returns - list of strings, representing responses
 */
function getResponsesFromForm(form) {
    const formData = new FormData(form);

    var responses = [];
    for (const [key, value] of formData) {
        responses[parseInt(key) - 1] = value;
    }
    return responses;
}